Foundation/FileManager.swift

```diff
@@ -509,28 +509,172 @@ open class FileManager : NSObject {
+                    times.deallocate()
+                }
+                if utimensat(AT_FDCWD, pathPtr, times, AT_SYMLINK_NOFOLLOW) < 0 {
+                    throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno))
+                }
+            }
+        } else {
+            let acsTimeval = toTimeval(access ?? curAccess ?? Date())
+            let modTimeval = toTimeval(modification ?? curModified ?? Date())
+            let createTimeval = creation.map(toTimeval)
+            
+            try _fileSystemRepresentation(withPath: path) { (pathPtr) in
+                if let createTimeval = createTimeval, (creation ?? .distantPast) > (curCreate ?? .distantPast) {
+                    // Setting creation time acccording to [FreeBSD manual](https://www.freebsd.org/cgi/man.cgi?query=utimes&sektion=2)
+                    let times = UnsafeMutablePointer<timeval>.allocate(capacity: 2)
+                    times.initialize(to: acsTimeval)
+                    times.advanced(by: 1).initialize(to: createTimeval)
+                    defer {
+                        times.deinitialize(count: 2)
+                        times.deallocate()
+                    }
+                    if lutimes(pathPtr, times) < 0 {
+                        throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno))
+                    }
+                }
+                
+                let times = UnsafeMutablePointer<timeval>.allocate(capacity: 2)
+                times.initialize(to: acsTimeval)
+                times.advanced(by: 1).initialize(to: modTimeval)
+                defer {
+                    times.deinitialize(count: 2)
+                    times.deallocate()
+                }
+                if lutimes(pathPtr, times) < 0 {
+                    throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno))
+                }
+            }
+        }
+        
+    }
+    
     /* setAttributes:ofItemAtPath:error: returns YES when the attributes specified in the 'attributes' dictionary are set successfully on the item specified by 'path'. If this method returns NO, a presentable NSError will be provided by-reference in the 'error' parameter. If no error is required, you may pass 'nil' for the error.
      
         This method replaces changeFileAttributes:atPath:.
      */
     open func setAttributes(_ attributes: [FileAttributeKey : Any], ofItemAtPath path: String) throws {
-        for attribute in attributes.keys {
-            if attribute == .posixPermissions {
-                guard let number = attributes[attribute] as? NSNumber else {
-                    fatalError("Can't set file permissions to \(attributes[attribute] as Any?)")
+        for (key, value) in attributes {
+            switch key {
+            case .creationDate:
+                guard let date = value as? Date else {
+                    fatalError("Can't set file modification date to \(value as Any?)")
+                }
```

Comment:
Since this function `throws` it is better to throw an error rather than call `fatalError()`

Reply:
I've duplicated Darwin's version behavior. It does crash with `NSException` if you pass another type, like string.

Reply:
In that case it might be better to use `as!` instead of the `as?` / `fatalError()` as it will have the same outcome and should simplify things a bit.

Reply:
These messages are exactly the NSException's messages which Darwin throws. That means we would have just a cryptic crash when final user uses compiled version of SF.

Reply:
Ah, that makes sense. However in the `setResourceValues` method `fatalError()` is called without a message so those ones could be changed.

Reply:
Ok done

Reply:
Agreed here that passing the wrong type in is a developer error worthy of trapping. I think the message could be clearer about the problem -- it's not important to me to maintain exactly the same string for a fatal error as an `NSException` on Darwin.

Reply:
@parkera I'm not a native speaker. I would appreciate if you have alternative sentences to replace.